osbuild/util/chroot.py

```diff
@@ -0,0 +1,41 @@
+import os
+import subprocess
+
+
+class Chroot:
+    """
+    Sets up mounts for the virtual filesystems inside a root tree, preparing it for running commands using chroot. This
+    should be used whenever a stage needs to run a command against the root tree but doesn't support a --root option or
+    similar.
+    Cleans up mounts when done.
+
+    This mounts /proc, /dev, and /sys.
+    """
+
+    def __init__(self, root: str):
+        self.root = root
+
+    def __enter__(self):
+        for d in ["/proc", "/dev", "/sys"]:
+            if not os.path.exists(self.root + d):
+                print(f"Making missing chroot directory: {d}")
+                os.makedirs(self.root + d)
+
+        subprocess.check_call(["/usr/bin/mount",
+                               "-t", "proc",
+                               "-o", "nosuid,noexec,nodev",
+                               "proc", f"{self.root}/proc"])
+        subprocess.check_call(["/usr/bin/mount",
+                               "-t", "devtmpfs",
+                               "-o", "mode=0755,noexec,nosuid,strictatime",
+                               "devtmpfs", f"{self.root}/dev"])
+        subprocess.check_call(["/usr/bin/mount",
+                               "-t", "sysfs",
+                               "-o", "nosuid,noexec,nodev",
+                               "sysfs", f"{self.root}/sys"])
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, tracebk):
+        for d in ["/proc", "/dev", "/sys"]:
+            subprocess.check_call(["/usr/bin/umount", self.root + d])
```

Comment:
(super nitpick) we should probably try to umount all even if one fails, right now if proc would fail to umount we would leave /dev, /sys hanging as well

Reply:
Done.

Reply:
https://github.com/achilleas-k/osbuild/blob/25ee837b35bf20cdef750faab7646697a738c510/stages/test/test_dracut.py#L25
This dracut test is failing now with this change but I'm not really sure what it's testing.

Reply:
So, it mocks the `subprocess.run` call and then checks its arguments. I don't remember the previous version, but if it was using `subprocess.run`, it would increase the number of calls accordingly, to 4. The rest of the test should be the same, since the dracut call would be still the first.

Reply:
ah, I see.  I thought maybe it was something related to the output (stdout changes), but yes it makes sense now that it's counting 4 calls to `subprocess.run()` because I changed all the `umount` calls from `check_call()` to `run()`.

Reply:
I'll prep the follow-up and include this.